sklearn/preprocessing/tests/test_imputation.py

```diff
@@ -345,3 +346,100 @@ def test_imputation_copy():
 
     # Note: If X is sparse and if missing_values=0, then a (dense) copy of X is
     # made, even if copy=False.
+
+def test_imputation_knn():
+    # Test imputation using knn strategy.
+    X = np.array([
+        [np.nan, -1,  0,  5],
+        [0,  2, -1,  3],
+        [-1,  -1,  0, 5],
+        [-1,  2,  3,  7],
+    ])
+
+    X2 = np.array([
+        [np.nan, -1,  0,  np.nan],
+        [0,  2, -1,  3],
+        [-1,  -1,  0, 5],
+        [-1,  2,  3,  7],
+    ])
+
+    X3 = np.array([
+        [np.nan, -1,  0,  5],
+        [0,  np.nan, -1,  3],
+        [-1,  -1,  np.nan, 5],
+        [-1,  2,  3,  np.nan],
+    ])
+
+    X4 = np.array([
+        [np.nan, -1,  0,  5],
+        [np.nan,  2, -1,  3],
+        [-1,  -1,  0, 5],
+        [0,  2,  -1,  6],
+    ])
+
+    X5 = np.array([
+        [999, -1,  0,  5],
+        [0,  2, -1,  3],
+        [-1,  -1,  0, 5],
+        [-1,  2,  3,  7],
+    ])
+
+    X_true_1 = np.array([
+        [-1, -1,  0,  5],
+        [0,  2, -1,  3],
+        [-1,  -1,  0, 5],
+        [-1,  2,  3,  7],
+    ])
+
+    X_true_2 = np.array([
+        [-0.5, -1,  0,  5],
```

Comment:
silly question: why isn't the last entry 4? The two nearest neighbors are X[1] and X[2], right?